libs/core/synchronization/include/hpx/synchronization/condition_variable.hpp

```diff
@@ -43,43 +105,109 @@ namespace hpx {
+         *                      the \a wait functions in a loop or using the
+         *                      overloads of \a wait, \a wait_for, or
+         *                      \a wait_until that take a predicate. end note]
+         *
+         * IOW, \a ~condition_variable() can execute before a signaled thread
+         * returns from a wait. If this happens with \a condition_variable,
+         * that waiting thread will attempt to lock the destructed mutex.
+         * To fix this, there must be shared ownership of the data members
+         * FIXME(bhumit->hkaiser) shouldn't this be condition_variable_any?
+         * between the \a condition_variable_any object and the member
+         * functions \a wait (\a wait_for, etc.).
+         */
         ~condition_variable() = default;
 
+        /**
+         * \brief If any threads are waiting on \a *this, calling \a notify_one
+         * unblocks one of the waiting threads. 
+         * 
+         * FIXME(bhumit->hkaiser) is this correct?
+         * \param ec Used to hold error code value originted during the
+         * operation. Defaults to \a throws -- A special 'throw on error'
+         * \a error_code.
+         *
+         * \returns \a notify_one returns \a void.
+         */
         void notify_one(error_code& ec = throws)
         {
             std::unique_lock<mutex_type> l(data_->mtx_);
             data_->cond_.notify_one(HPX_MOVE(l), ec);
         }
 
+        /**
+         * \brief Unblocks all threads currently waiting for \a *this. 
+         * 
+         * FIXME(bhumit->hkaiser) is this correct?
+         * \param ec Used to hold error code value originted during the
+         * operation. Defaults to \a throws -- A special 'throw on error'
+         * \a error_code.
+         *
+         * \returns \a notify_all returns \a void.
+         */
         void notify_all(error_code& ec = throws)
         {
             std::unique_lock<mutex_type> l(data_->mtx_);
             data_->cond_.notify_all(HPX_MOVE(l), ec);
         }
 
+        /**
+         * \brief \a wait causes the current thread to block until the
+         * condition variable is notified or a spurious wakeup occurs,
+         * optionally looping until some predicate is satisfied
+         * \a (bool(pred())==true).
+         *
+         * Atomically unlocks lock, blocks the current executing thread, and
+         * adds it to the list of threads waiting on \a *this. The thread will
+         * be unblocked when \a notify_all() or \a notify_one() is executed.
+         * It may also be unblocked spuriously. When unblocked, regardless of
+         * the reason, lock is reacquired and wait exits.
+         * 
+         * If this function fails to meet the postconditions
+         * (\a lock.owns_lock()==true and \a lock.mutex() is locked by the
+         * FIXME(bhumit->hkaiser) what does hpx call?
+         * calling thread), \a std::terminate is called. For example, this
```

Comment:
@hkaiser does hpx call `std::terminate` as well?

Reply:
Good question. I don't think we handle this case ATM. We should revisit the code.